### DISCUSSION TITLE
Building detectron2 on M1/M2 Macs

### DISCUSSION THREAD

**Discussion by jessecanada** [facebookresearch/detectron2#4336] [Show and tell]
Hi all,

With the new pytorch support for Apple Silicon, I was eager to try and run my detectron2 projects on my M1 Mac. I successfully used the following recipe to install detectron2. 

**Important**: The grpcio package is required to successfully build detectron2 from source, but it will not install right out of the box on Apple Silicon. To get circumvent this issue, we need to export the following environmental variables. [View source.](https://stackoverflow.com/questions/66640705/how-can-i-install-grpcio-on-an-apple-m1-silicon-laptop)
```export GRPC_PYTHON_BUILD_SYSTEM_OPENSSL=1```
```export GRPC_PYTHON_BUILD_SYSTEM_ZLIB=1```

After that, we are good to go.

1. Create a conda virtual environment
2. ***critical*** Install pytorch nightly build. However, at the time of writing (6/14/2022), torchaudio is not available yet as a prebuild. Thus, we use this command: ```pip install --pre torch torchvision torchaudio --extra-index-url https://download.pytorch.org/whl/nightly/cpu```
3. Run ```uname -m``` in terminal to check your archflag (should be "arm64")
4. *critical* Build detectron2 from source: ```CC=clang CXX=clang++ ARCHFLAGS="-arch arm64” python -m pip install 'git+https://github.com/facebookresearch/detectron2.git'# (add --user if you don't have permission)

**Comment by kimile599** [comment]
good to know that

**Comment by ccerhan** [comment]
For step 2, `torch` now can be installed like this:

`conda install pytorch==1.13.1 torchvision==0.14.1 -c pytorch`

Before step 4, `pycocotools` should be installed correctly for arm64:

`conda install pycocotools -c conda-forge`

Then install `detectron2`:

`CC=clang CXX=clang++ ARCHFLAGS="-arch arm64" python -m pip install 'git+https://github.com/facebookresearch/detectron2.git'`

**Reply by mayank1101** [reply]
It doesn't work

**Comment by dhruvsojitra76** [comment]
Detectron2 error: 

NotImplementedError: The operator 'aten::_foreach_add.List' is not currently implemented for the MPS device.

Is there any way i can utilize my MPS while traning on coustom dataset ?

is MMdetection competable with MPS?

**Reply by ianujv4231** [reply]
BRO SAME ERROR, HOW DID U RESOLVE IT ??? 

FOR THE TRAINING SEGMENT : 
trainer = DefaultTrainer(cfg)
trainer.resume_or_load(resume=False)  # Start from scratch
trainer.train()

**Comment by valentinamisi** [comment]
Regarding step 3, I get this: x86_64, does this change step 4? Step 4 is not working for me.

**Reply by valentinamisi** [reply]
I tried this: CC=clang CXX=clang++ ARCHFLAGS="-arch x86_64” python -m pip install 'git+https://github.com/facebookresearch/detectron2.git' 

and I get the following error:  

fatal error: too many errors emitted, stopping now [-ferror-limit=]
      20 errors generated.
      error: command '/usr/bin/clang' failed with exit code 1
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for detectron2
  Running setup.py clean for detectron2
Failed to build detectron2
ERROR: Could not build wheels for detectron2, which is required to install pyproject.toml-based projects

**Reply by OBW93** [reply]
I get the same error message with arm64 and using Python 3.9.4 with Sonoma 14.3 and a M1. Could you solve this problem?

**Reply by EmmaMeeus** [reply]
Same here, but with Python 3.10.0 / M1 / Sonoma 14.1.2.

**Reply by ezalorsara** [reply]
i got the same error

**Reply by tuokcalb** [reply]
I'm still stuck, did anyone ever get this working? I may just give up at this point.

**Reply by aguevaracaprio** [reply]
@valentinamisi you should be using ARCHFLAGS="-arch arm64” as M chips are based on ARM architecture. If you are on an Intel chip the x86 should be good. Was anyone able to make it work using M chips GPU support?

**Comment by peter65374** [comment]
it works on my Mac M2 chip with macOS14.2.1 Sonoma. That's cool.

**Comment by attentionmech** [comment]
wow, it works in my macOS 15.0 Sequoia. Thanks a tonn.

**Comment by LatteCoconut** [comment]
I failed to install it in pyenv using python3.10.
but I succeeded in install it with python3.9 in conda.
I don't understand

**Comment by gbmerrall** [comment]
I got it compiling with the following on an M2

```
CPATH=/Library/Developer/CommandLineTools/SDKs/MacOSX15.2.sdk/usr/include/c++/v1/ CC=clang CXX=clang++ ARCHFLAGS="-arch arm64" python -m pip install 'git+https://github.com/facebookresearch/detectron2.git'
```

The specific include directory may be different for you based on the version of CommandLineTools you  so check that before trying yourself.

**Reply by JoostvDoorn** [reply]
It seems you don't need to refer to the exact version, since there is a symlink to the latest version. The following worked for me.

```sh
CPATH=/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1 CC=clang CXX=clang++ ARCHFLAGS="-arch arm64" python -m pip install 'git+https://github.com/facebookresearch/detectron2.git'
```

**Reply by Benji-create** [reply]
i got it to install wtih

```
SDK_PATH=$(xcrun --show-sdk-path)
CPATH="${SDK_PATH}/usr/include/c++/v1/" CC=clang CXX=clang++ ARCHFLAGS="-arch arm64" python -m pip install 'git+https://github.com/facebookresearch/detectron2.git'
```

**Comment by Athleteg** [comment]
This is a workable solution for me:
% git clone https://github.com/facebookresearch/detectron2.git
% cd ~/detectron2
% CC=clang CXX=clang++ ARCHFLAGS="-arch arm64" pip3 install -e . --no-build-isolation
